src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -681,17 +681,28 @@ private void flush(boolean commitTranslog, boolean force, boolean waitIfOngoing)
                 throw new FlushNotAllowedEngineException(shardId, "recovery is in progress, flush with committing translog is not allowed");
             }
         }
-        int currentFlushing = flushing.incrementAndGet();
-        if (currentFlushing > 1 && waitIfOngoing == false) {
-            flushing.decrementAndGet();
-            throw new FlushNotAllowedEngineException(shardId, "already flushing...");
-        }
-
-        flushLock.lock();
-        try {
-             if (commitTranslog) {
-                try (ReleasableLock _ = readLock.acquire()) {
-                    ensureOpen();
+        /*
+         * Unfortunately the lock order is important here. We have to acquire the readlock fist otherwise
+         * if we are flushing at the end of the recovery while holding the write lock we can deadlock if:
+         *  Thread 1: flushes via API and gets the flush lock but blocks on the readlock since Thread 2 has the writeLock
+         *  Thread 2: flushes at the end of the recovery holding the writeLock and blocks on the flushLock owned by Thread 1
+         */
+        try (ReleasableLock _ = readLock.acquire()) {
+            final int currentFlushing = flushing.incrementAndGet();
+            if (currentFlushing > 1 && waitIfOngoing == false) {
+                flushing.decrementAndGet();
+                throw new FlushNotAllowedEngineException(shardId, "already flushing...");
+            }
+            if (waitIfOngoing == false) {
+                if (flushLock.tryLock() == false) {
+                    flushing.decrementAndGet();
```

Comment:
I think we need to throw an exception here, just as we do above with the currentFlushing check.